lockfile/src/spdx.rs

```diff
@@ -76,62 +79,84 @@ fn type_from_url(url: &str) -> anyhow::Result<PackageType> {
     }
 }
 
+fn from_purl(pkg_url: &str, pkg_info: &PackageInformation) -> anyhow::Result<Package> {
+    let purl = PackageUrl::from_str(pkg_url)?;
+
+    let package_type = PackageType::from_str(purl.ty())
+        .or_else(|_| type_from_url(&pkg_info.download_location))
+        .context(UnknownEcosystem)?;
+
+    let name = match (package_type, purl.namespace()) {
+        (PackageType::Maven, Some(ns)) => format!("{}:{}", ns, purl.name()),
+        (PackageType::Npm | PackageType::Golang, Some(ns)) => format!("{}/{}", ns, purl.name()),
+        _ => purl.name().into(),
+    };
+
+    let pkg_version = pkg_info
+        .version_info
+        .as_ref()
+        .ok_or(purl.version())
+        .map_err(|_| anyhow!("No version found for `{}`", pkg_info.name))?;
+
+    let version = purl
+        .qualifiers()
+        .iter()
+        .find_map(|(key, value)| match key.as_ref() {
+            "repository_url" => Some(PackageVersion::ThirdParty(ThirdPartyVersion {
+                version: pkg_version.clone(),
+                registry: value.to_string(),
+            })),
+            "download_url" => Some(PackageVersion::DownloadUrl(value.to_string())),
+            "vcs_url" => {
+                if value.as_ref().starts_with("git+") {
+                    Some(PackageVersion::Git(value.to_string()))
+                } else {
+                    None
+                }
+            },
+            _ => None,
+        })
+        .unwrap_or(PackageVersion::FirstParty(pkg_version.into()));
+
+    Ok(Package { name, version, package_type })
+}
+
+fn from_locator(registry: &str, locator: &str) -> anyhow::Result<Package> {
+    lazy_static! {
+        static ref RE: Regex = Regex::new(r"^(([^:]*:)*[^@:]*@?[^@:/]*)[:@/]([^:]+)$").unwrap();
+    }
+
+    let package_type = PackageType::from_str(registry).map_err(|_| UnknownEcosystem)?;
+    let decoded_locator = decode(locator).unwrap();
```

Comment:
Two questions:

1. Do all SPDX tools use URL encoding? I don't see it in the spec... But the spec also implies that `npm` package names won't contain a `@` (which they often do...)
2. It seems like the URL decode should happen **after** the regex and not before... For example, `%40npmcli/fs@2.1.2` is easier to parse as a locator than `@npmcli/fs@2.1.2`.

Reply:
1. I've seen the PURL ones use encoding. I just took a defensive approach.
2. Moved to just decode the name, but I can remove entirely